Issue #3018 Make paths in feature files OS independent

agajdosi opened the pull request:
Fixes #3018 

Steps to test the pull request

  1. make integration GODOG_OPTS=-tags=cmd-config

agajdosi commented:
retest this please

agajdosi commented on test/integration/features/cmd-config.feature:
```diff
@@ -176,15 +176,7 @@ user defined options which changes default behaviour of Minishift.
     | iso-url         | centos                                                                                           | centos                                                                                          |
-       
-  @quick
-  Scenario: Setting iso-url key to a non existing path
```
@anjannath Isn't it covered in unit tests? Not sure now.

anjannath replied:
@agajdosi Ohh.. yes this is extensively covered in unit test, sorry i forgot.

LalatenduMohanty commented on test/integration/features/cmd-config.feature:
```diff
@@ -175,15 +175,7 @@ user defined options which changes default behaviour of Minishift.
-     Then stderr should contain
-      """
-      'file:///home/Downloads/my_handmade_centos.iso' file is not present
-      """
```
@agajdosi If we remove this test then we wont be testing file URI for ISOs at all. So we can not remove this . However I am wondering if godog provides anyway to change a command for specific operating system.

anjannath replied:
@LalatenduMohanty Our unit test does cover this scenario already.

LalatenduMohanty replied:
@anjannath We can not keep track of individual unit tests vs integration tests. What happens when because of some reason the unit test does not catch the issue. I remember we had once missed to catch the bug with respect to file URI and personally do not want to miss issues like that.

agajdosi replied:
@LalatenduMohanty ack, so will try to figure out how to implement this nicely to be independent of platform. Unfortunately godog does not provide anything than tags which could be used. So probably small custom step will be needed, will check that.

gbraad commented:
is this a PR or an issue?

the title is confusing as it suggest this is just the issue report.
If thre is a fix, please also describe this in the title of the PR.
Else it might get ignored

agajdosi commented:
Need to test it on Windows, will then remove the [WIP].

agajdosi commented:
@gbraad Was not sure what will be the actual solution, so kept the complete issue name as the name of this PR. Naming is fixed now.

@LalatenduMohanty @anjannath PR is ready now for review. I have tested on Linux and Windows and it works fine. I have added support for scenario variables to be global, no longer they are usable on only some steps. Also I have added default scenario variable which contains path to test directory (`TEST_DIR`) - on Linux it is something like: `file:///home/agajdosi/...`, on Windows then `file://C:/Users/agajdosi/...`. In config file I then just use `When executing "minishift config set iso-url $(TEST_DIR)/non-existing.iso" fails`.

gbraad commented on test/integration/testsuite/prepare.go:
```diff
@@ -111,6 +111,10 @@ func SetMinishiftHomeEnv(path string) {
 }
 
+func setDefaultScenarioVariables() {
+	util.SetVariable("TEST_DIR", "file://"+testDir)
```
defaults ;-)

gbraad commented on test/integration/testsuite/testsuite.go:
```diff
@@ -355,10 +343,31 @@ func FeatureContext(s *godog.Suite) {
 	})
 
+	s.BeforeStep(func(this *gherkin.Step) {
+		this.Text = util.ProcessVariables(this.Text)
```
👍

praveenkumar commented on Gopkg.toml:
```diff
@@ -25,7 +25,7 @@ ignored = ["github.com/Sirupsen/logrus"]
 [[constraint]]
   name = "github.com/DATA-DOG/godog"
-  version = "=0.7.6"
+  branch = "master"
```
We need to put a revision to make sure this will not make the change to lock file when someone run `ensure`.

praveenkumar commented:
Fixed revision and merged

agajdosi commented:
Thanks for merging it guys :)
